{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing promoter data\n",
    "\n",
    "Last updated: January 28, 2019\n",
    "\n",
    "Created by: Miguel A. Alcantar\n",
    "\n",
    "\n",
    "This notebook parses through data from various promoter databases and concatenates the data into a dataframe.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# package versions \n",
    "\n",
    "print(\"Pandas: v\" + pd.__version__) \n",
    "print(\"Numpy: v\" + np.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### functions for filtering and QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_range(promoter_sequence_df):\n",
    "    \n",
    "    \"\"\"\n",
    "    calculate range of provided promoter sequences, with respect to the TSS. \n",
    "    this is done based on the delimiter '/', which surrounds the TSS (+1).\n",
    "    \n",
    "    inputs: dataframe containing promoter sequences. sequences must be under column\n",
    "    named \"range (with respect to TSS)\".\n",
    "    \n",
    "    outputs: dataframe containing promoter sequences with updated \"range (with respect to TSS)\".\n",
    "    \"\"\"\n",
    "    \n",
    "    # loop through dataframe and update range if there is a '/' marking the TSS\n",
    "    for idx, sequence in promoter_sequence_df['DNA sequence'].iteritems():\n",
    "        \n",
    "        if '/' in sequence:\n",
    "                seq_split = sequence.split('/')\n",
    "                TSS_upstream_len = str(len(seq_split[0]) - 1)\n",
    "                TSS_downstream_len = str(len(seq_split[2]) + 1)\n",
    "                range_string = '-' + TSS_upstream_len + ' to ' + TSS_downstream_len\n",
    "                promoter_sequence_df.loc[idx, 'range (with respect to TSS)'] = range_string\n",
    "                \n",
    "    return(promoter_sequence_df)\n",
    "\n",
    "def QC_DNA_sequences(promoter_sequence_df):\n",
    "    \n",
    "    \"\"\"\n",
    "    perform quality control (QC) and filter dna promoter sequences. QC involves the following:\n",
    "    i) remove unwanted char acters: line-break character '\\n', spaces, and brackets\n",
    "    ii) use the \"calc_range\" function to update missing ranges with a known TSS\n",
    "        '/' characters denoting TSS are removed after this step\n",
    "    iii) replace NaN elements with empty string \n",
    "    iv) removing sequences that don't contain canonical nucleotides: a, t, c, g\n",
    "    this function also prints out the number of sequences filtered\n",
    "    \n",
    "    inputs: dataframe containing promoter sequences. sequences must be under column\n",
    "    named \"range (with respect to TSS)\".\n",
    "    \n",
    "    outputs: dataframe that has undergone QC and filtering\n",
    "    \"\"\"\n",
    "    \n",
    "    original_sequence_number = promoter_sequence_df.shape[0]\n",
    "    \n",
    "    if list(promoter_sequence_df['database/source'])[0] == 'EPDnew':\n",
    "        db = 'EPDnew'\n",
    "    elif list(promoter_sequence_df['database/source'])[0] == 'RegulonDB':\n",
    "        db = 'RegulonDB'\n",
    "    elif list(promoter_sequence_df['database/source'])[0] == 'DBTBS':\n",
    "        db = 'DBTBS'\n",
    "    elif list(promoter_sequence_df['database/source'])[0] == 'Meyer et al 2019 (PMID: 30478458)':\n",
    "        db = 'bacterial inducible promoters'\n",
    "    else:\n",
    "        db = 'fungal inducible promoters'\n",
    "    \n",
    "    # remove unwanted characters from DNA sequences\n",
    "    sequences = list(promoter_sequence_df['DNA sequence'])\n",
    "    sequences = [str(seq).strip('\\n') for seq in sequences] # remove edge '\\n'\n",
    "    sequences = [seq.replace('\\n','') for seq in sequences] # remove other '\\n'\n",
    "    sequences = [seq.replace(' ','') for seq in sequences] # remove empty spaces\n",
    "    sequences = [seq.lower() for seq in sequences] # convert sequences to lowercase\n",
    "    sequences = [seq.replace('{', '').replace('}', '') for seq in sequences]\n",
    "    \n",
    "    promoter_sequence_df['DNA sequence'] = sequences\n",
    "    # update range (with respect to TSS)\n",
    "    promoter_sequence_df = calc_range(promoter_sequence_df)\n",
    "    sequences = [seq.replace('/', '') for seq in sequences]\n",
    "    \n",
    "    # replace old sequences with QC sequences\n",
    "    promoter_sequence_df['DNA sequence'] = sequences\n",
    "\n",
    "    # filter rows that don't only contain nucleotide sequence\n",
    "    promoter_sequence_df = promoter_sequence_df.fillna('').replace('NaN','').replace('None','')\n",
    "    promoter_sequence_df = promoter_sequence_df[(~promoter_sequence_df['DNA sequence'].str.contains('[^atcg]')) & (promoter_sequence_df['DNA sequence'].str.len() > 0)]\n",
    "    promoter_sequence_df.reset_index()\n",
    "    new_sequence_number = promoter_sequence_df.shape[0]\n",
    "    \n",
    "    # display number of filtered sequences\n",
    "    print(str(original_sequence_number-new_sequence_number) + ' sequences filtered from ' + db + ' dataset')\n",
    "    \n",
    "    return(promoter_sequence_df)\n",
    "\n",
    "def rename_promoter_df_columns(promoter_sequence_df, promoter_df_or_type):\n",
    "    \n",
    "    \"\"\"\n",
    "    standardize column names for dataframe containing promoter sequences and metadata. \n",
    "    new columns are also added, according the selected database\n",
    "    \n",
    "    inputs: dataframe containing promoter sequences and metedata, \n",
    "            string indicating promoter database or type. valid string inputs are: \"RegulonDB\", DBTBS\n",
    "                string input is NOT case-sensitive\n",
    "    \n",
    "    outputs: dataframe containing promoter sequences with updated column names\n",
    "    \"\"\"\n",
    "    \n",
    "    # standardized column names for RegulonDB dataframe\n",
    "    newregulonDB_columns_list = [\"RegulonDB ID\", \n",
    "                                \"promoter\",\n",
    "                                \"DNA strand\",\n",
    "                                \"location of TSS on genome\",\n",
    "                                \"sigma factor/motif\",\n",
    "                                \"DNA sequence\",\n",
    "                                \"evidence\",\n",
    "                                \"evidence confidence\"]\n",
    "    \n",
    "    # standardized column names for DBTBS dataframe\n",
    "    newDBTBS_columns_list = ['operon', \n",
    "                             'regulated gene', \n",
    "                             'absolute position', \n",
    "                             'range (with respect to TSS)',\n",
    "                             'DNA sequence', \n",
    "                             'experimental evidence', \n",
    "                             'sigma factor/motif']\n",
    "    \n",
    "    # standardize column names and add new columns, as needed\n",
    "    if promoter_df_or_type.lower() == 'regulondb':\n",
    "        \n",
    "        promoter_sequence_df_columns_dict = dict(zip(list(promoter_sequence_df.columns), newregulonDB_columns_list))\n",
    "        promoter_sequence_df = promoter_sequence_df.rename(columns=promoter_sequence_df_columns_dict)\n",
    "        promoter_sequence_df['database/source'] = \"RegulonDB\"\n",
    "        promoter_sequence_df['organism'] = 'e_coli'\n",
    "        promoter_sequence_df['range (with respect to TSS)'] = '-60 to +20'         \n",
    "    \n",
    "    elif promoter_df_or_type.lower() == 'dbtbs':\n",
    "\n",
    "        promoter_sequence_df_columns_dict = dict(zip(list(promoter_sequence_df.columns), newDBTBS_columns_list))\n",
    "        promoter_sequence_df = promoter_sequence_df.rename(columns=promoter_sequence_df_columns_dict)\n",
    "        promoter_sequence_df['organism'] = 'b. subtilis'\n",
    "        promoter_sequence_df['database/source'] = 'DBTBS'\n",
    "        \n",
    "        # standardize range syntax  \n",
    "        DBTBS_range_list = list(promoter_sequence_df['range (with respect to TSS)'])\n",
    "        range_new_list = []\n",
    "        for range_orig_list in DBTBS_range_list:\n",
    "            range_orig_splt = str(range_orig_list).split(':')\n",
    "            if len(range_orig_splt) > 1:\n",
    "                if int(range_orig_splt[0]) < 0:\n",
    "                    range_orig_lower_range = str(int(range_orig_splt[0]) + 1)\n",
    "                else:\n",
    "                    range_orig_lower_range = range_orig_splt[0]\n",
    "\n",
    "                range_orig_upper_range = range_orig_splt[1]\n",
    "                range_new = range_orig_lower_range + ' to ' + range_orig_upper_range\n",
    "            else:\n",
    "                range_new = range_orig_list\n",
    "            range_new_list.append(range_new)\n",
    "        promoter_sequence_df['range (with respect to TSS)'] = range_new_list\n",
    "#         DBTBS_range_list = list(promoter_sequence_df['range (with respect to TSS)'])\n",
    "#         DBTBS_range_list_renamed = [str(seq_range).replace(':', ' to ') for seq_range in DBTBS_range_list]\n",
    "#         promoter_sequence_df['range (with respect to TSS)'] = DBTBS_range_list_renamed\n",
    "    else:\n",
    "        print('Error! Please enter a valid promoter dataframe or type: \\'RegulonDB\\' or \\'DBTBS\\'')\n",
    "\n",
    "    return(promoter_sequence_df)\n",
    "    \n",
    "def reorganize_promoter_df_columns(promoter_sequence_df):\n",
    "    \n",
    "    \"\"\"\n",
    "    reorganize promoter dataframe columns such that they are consistent between databases.\n",
    "    this will involve dropping some unnecessary columns\n",
    "    \n",
    "    inputs: dataframe containing promoter sequences\n",
    "        this dataframe will ideally have been passed through the rename_promoter_df_columns function if needed\n",
    "        \n",
    "    outputs: dataframe with reordered columns  \n",
    "    \"\"\"\n",
    "    \n",
    "    columns_to_conserve = ['organism', 'database/source', 'DNA sequence', 'regulated gene', \n",
    "                           'range (with respect to TSS)', 'motifs', 'sigma factor/motif', 'inducer/repressor', 'promoter']\n",
    "    \n",
    "    columns_to_add = list(set(columns_to_conserve) - set(promoter_sequence_df.columns))\n",
    "    columns_to_drop = list(set(promoter_sequence_df.columns) - set(columns_to_conserve))\n",
    "    \n",
    "    # if new columns are added, populate with empty strings\n",
    "    for col_to_add in columns_to_add:\n",
    "        promoter_sequence_df[col_to_add] = ''\n",
    "        \n",
    "    promoter_sequence_df = promoter_sequence_df.drop(columns_to_drop, axis=1)\n",
    "    promoter_sequence_df = promoter_sequence_df[columns_to_conserve]\n",
    "    \n",
    "    return(promoter_sequence_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### functions for finding number of motifs in datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def EPDnew_motifs(EPDnew_promoters_df, save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    stratify EPDnew promoters by motif and organism\n",
    "    \n",
    "    inputs: EPDnew promoter dataframe, \n",
    "            boolean indicating whether to save csv\n",
    "            \n",
    "    outputs: dataframe containing matrix of organism x motif (nonTATA, TATA, Inr-pfa, nonInr-pfa, total)    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    organism_motifs_dict = dict()\n",
    "    organisms_list = list(set(EPDnew_promoters_df['organism']))\n",
    "    motifs_list = list(set(EPDnew_promoters_df['sigma factor/motif']))\n",
    "\n",
    "    # calculate number of promoters that were obtained for each organism (and further stratify by motif)\n",
    "    for motif in motifs_list:\n",
    "        motif_occurances_list = []\n",
    "        for organism in organisms_list:\n",
    "            organism_motifs_count = EPDnew_promoters_df[(EPDnew_promoters_df['sigma factor/motif'] == motif) & (EPDnew_promoters_df['organism'] == organism)].shape[0]\n",
    "            motif_occurances_list.append(organism_motifs_count)\n",
    "        organism_motifs_dict.update({motif: motif_occurances_list})\n",
    "    \n",
    "    # creating dataframe of organism x motif (a form that is useful for plotting)\n",
    "    motifs_occurances_df = pd.DataFrame.from_dict(organism_motifs_dict,orient='index').transpose()\n",
    "    \n",
    "    # replace underscors with periods for organism name\n",
    "    organisms_replace_underscore = [org.replace(\"_\",\". \") for org in organisms_list]\n",
    "    reindex_motifs_occuranges_dict = dict(zip(list(motifs_occurances_df.index), organisms_replace_underscore)) \n",
    "    motifs_occurances_df = motifs_occurances_df.rename(index = reindex_motifs_occuranges_dict)\n",
    "    \n",
    "    # find total occurance occurance of promoters for each organism and then organize by motif\n",
    "    motifs_occurances_df['total'] = motifs_occurances_df[list(motifs_occurances_df.columns)].sum(axis=1)\n",
    "    motifs_occurances_df = motifs_occurances_df.sort_values(by=\"total\")\n",
    "    motifs_occurances_df = motifs_occurances_df[['nonTATA', 'TATA','Inr-pfa', 'nonInr-pfa', 'total']]\n",
    "    \n",
    "    # find total number of each kind of motif\n",
    "    motifs_occurances_df.loc['total'] = np.nan\n",
    "    motifs_occurances_df.loc['total', 'nonTATA'] = motifs_occurances_df.loc[:, 'nonTATA'].sum()\n",
    "    motifs_occurances_df.loc['total', 'TATA'] = motifs_occurances_df.loc[:, 'TATA'].sum()\n",
    "    motifs_occurances_df.loc['total', 'Inr-pfa'] = motifs_occurances_df.loc[:, 'Inr-pfa'].sum()\n",
    "    motifs_occurances_df.loc['total', 'nonInr-pfa'] = motifs_occurances_df.loc[:, 'nonInr-pfa'].sum()\n",
    "    motifs_occurances_df.loc['total', 'total'] = motifs_occurances_df.loc[:, 'total'].sum()\n",
    "    \n",
    "    # convert dataframe datatype to integer \n",
    "    motifs_occurances_df = motifs_occurances_df.astype('int')\n",
    "    \n",
    "    if save_csv:\n",
    "        motifs_occurances_df.to_csv('../../data/parsed_promoter_data/20191125_EPDnew_motifs_occurances.csv')\n",
    "    \n",
    "    return(motifs_occurances_df)\n",
    "\n",
    "def RegulonDB_motifs(regulonDB_promoters, save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    stratify RegulonDB promoters by sigma factor that binds promoter\n",
    "    \n",
    "    inputs: RegulonDB promoter dataframe, \n",
    "            boolean indicating whether to save csv\n",
    "            \n",
    "    outputs: dataframe containing matrix of sigma factor x number of promoters\n",
    "    \"\"\"\n",
    "    \n",
    "    # define E. coli sigma factors\n",
    "    sigma_factors_Ecoli = ['unknown',\n",
    "                           'Sigma70',\n",
    "                           'Sigma54',\n",
    "                           'Sigma38',\n",
    "                           'Sigma32',\n",
    "                           'Sigma28',\n",
    "                           'Sigma24',\n",
    "                           'Sigma19']\n",
    "\n",
    "    # dictionary that will contain promoter sigma factor frequencies             \n",
    "    Ecoli_sigma_factor_count_dict = dict()\n",
    "\n",
    "    # determine counts for each sigma factor\n",
    "    for sigma_factor in sigma_factors_Ecoli:\n",
    "        \n",
    "        if sigma_factor == 'unknown':\n",
    "            SF_count = int(regulonDB_promoters[regulonDB_promoters['sigma factor/motif'].str.contains(sigma_factor)].shape[0]) + \\\n",
    "                        int(regulonDB_promoters[regulonDB_promoters['sigma factor/motif'] == ''].shape[0])\n",
    "        else:\n",
    "            SF_count = regulonDB_promoters[regulonDB_promoters['sigma factor/motif'].str.contains(sigma_factor)].shape[0]\n",
    "        \n",
    "        Ecoli_sigma_factor_count_dict.update({sigma_factor: SF_count})\n",
    "\n",
    "    # build dataframe containing sigma factor count information\n",
    "    Ecoli_sigma_factor_occurances = pd.DataFrame.from_dict(Ecoli_sigma_factor_count_dict,orient='index')\n",
    "    Ecoli_sigma_factor_occurances.loc['total',:] = regulonDB_promoters.shape[0]\n",
    "    Ecoli_sigma_factor_occurances = Ecoli_sigma_factor_occurances.rename(columns = {0:\"E. coli\"}).sort_values(by=\"E. coli\")\n",
    "    Ecoli_sigma_factor_occurances = Ecoli_sigma_factor_occurances.astype('int')\n",
    "    \n",
    "    if save_csv:\n",
    "        Ecoli_sigma_factor_occurances.to_csv('../../data/parsed_promoter_data/20191202_RegulonDB_sigma_factor_occurances.csv')\n",
    "        \n",
    "    return(Ecoli_sigma_factor_occurances)\n",
    "\n",
    "def DBTBS_motifs(DBTBS_promoters_df, save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    stratify DBTBS promoters by sigma factor that binds promoter\n",
    "    \n",
    "    \n",
    "    inputs: DBTBS promoter dataframe, \n",
    "            boolean indicating whether to save csv\n",
    "            \n",
    "    outputs: dataframe containing matrix of sigma factor x number of promoters\n",
    "    \"\"\"\n",
    "      \n",
    "    sigma_factors_Bsubtilis = list(set(DBTBS_promoters_df['sigma factor/motif']))\n",
    "    Bsubtilis_sigma_factor_count_dict = dict()\n",
    "\n",
    "    # loop through sigma factors to find counts of each sigma factor type\n",
    "    for sigma_factor in sigma_factors_Bsubtilis:\n",
    "        SF_count = DBTBS_promoters_df[DBTBS_promoters_df['sigma factor/motif'].str.contains(sigma_factor[0:5])].shape[0]\n",
    "        Bsubtilis_sigma_factor_count_dict.update({sigma_factor: SF_count})\n",
    "        \n",
    "    Bsubtilis_sigma_factor_occurances = pd.DataFrame.from_dict(Bsubtilis_sigma_factor_count_dict,orient='index')\n",
    "    Bsubtilis_sigma_factor_occurances.loc[\"total\",:] = DBTBS_promoters_df.shape[0]\n",
    "    Bsubtilis_sigma_factor_occurances = Bsubtilis_sigma_factor_occurances.rename(columns = {0:\"B. subtilis\"}).sort_values(by=\"B. subtilis\")\n",
    "    Bsubtilis_sigma_factor_occurances = Bsubtilis_sigma_factor_occurances.astype('int')\n",
    "\n",
    "    if save_csv:\n",
    "        Bsubtilis_sigma_factor_occurances.to_csv(\"../../data/parsed_promoter_data/20191129_DBTBS_sigma_factor_occurances.csv\")\n",
    "        \n",
    "    return(Bsubtilis_sigma_factor_occurances)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EPDnew\n",
    "\n",
    "### Contains sequences for 15 different eukaryotic species\n",
    "\n",
    "1. C. elegans (worm)\n",
    "2. D. melanogaster (fly)\n",
    "3. D. rerio (zebrafish)\n",
    "4. R. norbegicus (rat)\n",
    "5. Z. mays (maize)\n",
    "6. H. sapien (humans)\n",
    "7. G. gallus (chicken)\n",
    "8. S. pombe (fission yeast)\n",
    "9. A. thaliana (arabidopsis)\n",
    "10. M. mulatta (monkey)\n",
    "11. P. falciparum (malaria-causing parasite)\n",
    "12. M. musculus (mouse)\n",
    "13. S. cerevisiae (yeast)\n",
    "14. A. mellifera (bee)\n",
    "15. C. familiaris (dog)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_EPDnew(save_csv=False):\n",
    "    \"\"\"\n",
    "    parse through EPDnew database and apply QC / filtering\n",
    "    \n",
    "    inputs: boolean argument indicating whether csv files should be saved\n",
    "    \n",
    "    outputs: dataframe containing EPDnew promoters and metadata,\n",
    "            if save_csv, then csv file is saved to data directory\n",
    "    \n",
    "    \"\"\"\n",
    "        \n",
    "    # hard-coding directory containing data of interest\n",
    "    directory_str = '../../data/EPDnew'\n",
    "    directory = os.fsencode(directory_str)\n",
    "\n",
    "    # initializing lists that will contain information of interest\n",
    "    promoter_organism = []\n",
    "    promoter_database = []\n",
    "    promoter_sequences = []\n",
    "    promoter_gene = []\n",
    "    promoter_range = []\n",
    "    promoter_notes = []\n",
    "\n",
    "    # loop through all files \n",
    "    # each file contains promoters for a particular organism and motif (e.g., h. sapien TATA promoters)\n",
    "    for file in os.listdir(directory):\n",
    "\n",
    "        # establishing file path and name\n",
    "        filename = os.fsdecode(file)\n",
    "        filepath = directory_str + '/' + filename\n",
    "\n",
    "        # initializing gene number and sequence (the latter helps with appending sequences to list)\n",
    "        counter_gene = 0 \n",
    "        sequence = ''\n",
    "        \n",
    "        # ignore hidden files\n",
    "        if filename[0] != '.': \n",
    "            \n",
    "            with open(filepath) as fp:\n",
    "\n",
    "                line = fp.readline()\n",
    "\n",
    "                while line:\n",
    "\n",
    "                    # The '>' signifies the start of a new gene\n",
    "                    if line[0] == '>':\n",
    "\n",
    "                        # append last completed sequence if we've already been through 1 iteration\n",
    "                        if counter_gene >= 1:\n",
    "\n",
    "                            promoter_sequences.append(sequence)\n",
    "                            \n",
    "                        sequence = ''\n",
    "                        counter_gene += 1\n",
    "                        \n",
    "                        line_split = line.split(' ')\n",
    "\n",
    "                        # extracting gene regulated by the current promoter\n",
    "                        promoter_gene.append(line_split[1])\n",
    "\n",
    "                        # edge case where gene does not have a specified range\n",
    "                        try:\n",
    "                            range_index = line.split(' ').index('range')\n",
    "                            range_temp = line.split(' ')[range_index+2] + ' to ' + line.split(' ')[range_index+6].strip('.\\n')\n",
    "                            promoter_range.append(range_temp)\n",
    "                        except ValueError:\n",
    "                            promoter_range.append('')\n",
    "\n",
    "                        # append organism name, database where promoter was obtained, and notes (e.g., TATA or nonTATA promoter)\n",
    "                        promoter_organism.append(filename.split('_')[2] + '_' + filename.split('_')[3])\n",
    "                        promoter_database.append(filename.split('_')[1])\n",
    "                        promoter_notes.append(filename.split('_')[4].replace(' ',''))\n",
    "\n",
    "                    else: # keep appending nucleotides until we reach a new sequence\n",
    "                        sequence = sequence + line\n",
    "                    line = fp.readline()\n",
    "            promoter_sequences.append(sequence)\n",
    "\n",
    "        # organizing data into a dictionary\n",
    "        EPDnew_data_dict = {\n",
    "            'organism': promoter_organism,\n",
    "            'database/source': promoter_database,\n",
    "            'DNA sequence':promoter_sequences,\n",
    "            'regulated gene': promoter_gene,\n",
    "            'range (with respect to TSS)': promoter_range,\n",
    "            'sigma factor/motif': promoter_notes\n",
    "                            }\n",
    "\n",
    "    # organizing data into dataframe\n",
    "    EPDnew_promoters_df = pd.DataFrame.from_dict(EPDnew_data_dict,orient='index').transpose()\n",
    "    EPDnew_promoters_df = QC_DNA_sequences(EPDnew_promoters_df) \n",
    "    \n",
    "    # save as csv \n",
    "    if save_csv:\n",
    "        EPDnew_promoters_df.to_csv('../../data/parsed_promoter_data/20191125_EPDnew_promoters.csv')\n",
    "    \n",
    "    return(EPDnew_promoters_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating stacked barplot showing number of promoters from each organism and their specific motifs\n",
    "def plot_EPDnew_data(motifs_occurances_df, save_plot=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    create bar plot showing EPDnew promoters stratified by organism and motif\n",
    "    \n",
    "    # inputs: dataframe containing EPDnew organisms x motifs (the output of EPDnew_motifs function)\n",
    "    \n",
    "    # outputs: if save_plot, then a bar plot is saved\n",
    "    \"\"\"\n",
    "    \n",
    "    dark_blue = '#4D98FA'\n",
    "    dark_yellow = '#FAAE65'\n",
    "    light_blue = '#60B5E6'\n",
    "    purple = '#CC7DAA'\n",
    "    plot_colors = [dark_blue, dark_yellow, light_blue, purple]\n",
    "\n",
    "    # create horizontal barplot that specifies promoter by organism and motif\n",
    "    motifs_occurances_df.iloc[:,:-1].plot.barh(stacked=True, color = plot_colors ,figsize=(10,7))\n",
    "    plt.xlabel('Count', size=24,fontname=\"Arial\"),\n",
    "    plt.ylabel('Organism', size=24,fontname=\"Arial\")\n",
    "    plt.title('Promoter counts by organism and motif',fontname=\"Arial\", size=24)\n",
    "    plt.xticks(fontsize = 16,fontname=\"Arial\") \n",
    "    plt.yticks(fontsize = 16,fontname=\"Arial\")\n",
    "    plt.legend(loc='lower right', prop={'size': 16})\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if save_plot:\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_EPDnew_promoters_by_organisms_and_motifs.pdf')\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_EPDnew_promoters_by_organisms_and_motifs.png', dpi=400)\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_EPDnew_promoters_by_organisms_and_motifs.svg')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run EPDnew functions\n",
    "EPDnew_promoters_df = parse_EPDnew(save_csv=False)\n",
    "motifs_occurances_df = EPDnew_motifs(EPDnew_promoters_df)\n",
    "plot_EPDnew_data(motifs_occurances_df, save_plot=False) # plot EPDnew data\n",
    "\n",
    "# edit EPDnew dataframe such that it is compatible for concatenation\n",
    "EPDnew_promoters_df_to_concat = reorganize_promoter_df_columns(EPDnew_promoters_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RegulonDB\n",
    "\n",
    "### Contains sequences for different E. coli polymerase sigma factors (70, 54, 38, 32, 28, 24, 19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_RegulonDB(save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    parse through RegulonDB database and apply QC / filtering\n",
    "    \n",
    "    inputs: boolean argument indicating whether csv files should be saved\n",
    "    \n",
    "    outputs: dataframe containing RegulonDB promoters and metadata,\n",
    "            if save_csv, then csv file is saved to data directory\n",
    "    \n",
    "    \"\"\"\n",
    "        \n",
    "    # Download regulon DB promoter database and organize dataframe\n",
    "    regulonDB_promoters_df = pd.read_csv('../../data/RegulonDB/20191127_PromoterSet.csv', header=36)\n",
    "    regulonDB_promoters_df = rename_promoter_df_columns(regulonDB_promoters_df,promoter_df_or_type='RegulonDB')\n",
    "    regulonDB_promoters_df = QC_DNA_sequences(regulonDB_promoters_df)\n",
    "\n",
    "    if save_csv:\n",
    "        regulonDB_promoters_df.to_csv('../../data/parsed_promoter_data/20191202_RegulonDB_promoters.csv')\n",
    "    \n",
    "    return(regulonDB_promoters_df)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_RegulonDB_data(Ecoli_sigma_factor_occurances, save_plot=False):   \n",
    "        \n",
    "    \"\"\"\n",
    "    create bar plot showing RegulonDB promoters stratified by cognate sigma factor\n",
    "    \n",
    "    # inputs: dataframe containing RegulonDB organisms x motifs (the output of RegulonDB_motifs function)\n",
    "    \n",
    "    # outputs: if save_plot, then a bar plot is saved\n",
    "    \"\"\"\n",
    "    \n",
    "    # plot promoter counts by sigma factor\n",
    "    Ecoli_sigma_factor_occurances.astype(int).plot.barh(figsize=(10,7), color = '#C4C4C4')\n",
    "    plt.xlabel('Count', size=24,fontname=\"Arial\")\n",
    "    plt.ylabel('E. coli sigma factor', size=24,fontname=\"Arial\")\n",
    "    plt.title('E. coli promoter counts by sigma factor', size=24,fontname=\"Arial\")\n",
    "    plt.xticks(fontsize = 16,fontname=\"Arial\") # work on current fig\n",
    "    plt.yticks(fontsize = 16,fontname=\"Arial\") # work on current fig\n",
    "    plt.legend(loc='lower right', prop={'size': 16})\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if save_plot:\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_RegulonDB_promoters_by_ecoli_sigmaFactor.pdf')\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_RegulonDB_promoters_by_ecoli_sigmaFactor.png', dpi=400)\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_RegulonDB_promoters_by_ecoli_sigmaFactor.svg')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run RegulonDB functions\n",
    "regulonDB_promoters_df = parse_RegulonDB(save_csv=False)\n",
    "Ecoli_sigma_factor_occurances = RegulonDB_motifs(regulonDB_promoters_df, save_csv=False)\n",
    "plot_RegulonDB_data(Ecoli_sigma_factor_occurances, save_plot=False)\n",
    "\n",
    "# edit RegulongDB dataframe such that it is compatible for concatenation\n",
    "regulonDB_promoters_df_to_concat = reorganize_promoter_df_columns(regulonDB_promoters_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DBTBS\n",
    "### Contains sequences for different B. subtilis polymerase sigma factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_DBTBS(save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    parse through DBTBS database and apply QC / filtering\n",
    "    \n",
    "    inputs: boolean argument indicating whether csv files should be saved\n",
    "    \n",
    "    outputs: dataframe containing DBTBS promoters and metadata,\n",
    "            if save_csv, then csv file is saved to data directory\n",
    "    \n",
    "    \"\"\"    \n",
    "    \n",
    "    # merge Bacillus subtilus promoter excel sheets into one comprehensive dataframe\n",
    "    DBTBS_all_promoter_sheets_df = pd.read_excel('../../data/DBTBS/20191129_DBTBS_promoter_sequences.xlsx', sheet_name=None, ignore_index=True)\n",
    "    DBTBS_promoters_df = pd.concat(DBTBS_all_promoter_sheets_df.values()).fillna('')# #dropna(thresh=1)\n",
    "    DBTBS_promoters_df.reset_index(drop=True, inplace=True)\n",
    "    DBTBS_promoters_df = rename_promoter_df_columns(DBTBS_promoters_df, promoter_df_or_type='DBTBS')\n",
    "    \n",
    "    DBTBS_promoters_df = QC_DNA_sequences(DBTBS_promoters_df)\n",
    "    \n",
    "    # save to csv\n",
    "    if save_csv:\n",
    "        DBTBS_promoters_df.to_csv(\"../../data/parsed_promoter_data/20191129_DBTBS_promoter_sequences.csv\")\n",
    "        \n",
    "    return(DBTBS_promoters_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_DBTBS_data(Bsubtilis_sigma_factor_occurances, save_plot=False):\n",
    "\n",
    "    \"\"\"\n",
    "    create bar plot showing DBTBS promoters stratified by cognate sigma factor\n",
    "    \n",
    "    # inputs: dataframe containing DBTBS organisms x motifs (the output of DBTBS_motifs function)\n",
    "    \n",
    "    # outputs: if save_plot, then a bar plot is saved\n",
    "    \"\"\"\n",
    "    \n",
    "    Bsubtilis_sigma_factor_occurances.astype(int).plot.barh(figsize=(10,7), color = '#6E8DAB')\n",
    "    plt.xlabel('Count', size=24,fontname=\"Arial\"),\n",
    "    plt.ylabel('B. subtilis sigma factor', size=24,fontname=\"Arial\")\n",
    "    plt.title('B. subtilis promoter counts by sigma factor', size=24,fontname=\"Arial\")\n",
    "    plt.xticks(fontsize = 16) # work on current fig\n",
    "    plt.yticks(fontsize = 16) # work on current fig\n",
    "    plt.legend(loc='lower right', prop={'size': 16})\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if save_plot:\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_DBTBS_promoters_by_Bsubtilis_sigmaFactor.pdf')\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_DBTBS_promoters_by_Bsubtilis_sigmaFactor.png', dpi=400)\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191127_DBTBS_promoters_by_Bsubtilis_sigmaFactor.svg')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run DBTBS functions\n",
    "DBTBS_promoters_df = parse_DBTBS(save_csv=False) #Bsubtilis_sigma_factor_occurances\n",
    "Bsubtilis_sigma_factor_occurances = DBTBS_motifs(DBTBS_promoters_df, save_csv=False)\n",
    "plot_DBTBS_data(Bsubtilis_sigma_factor_occurances, save_plot=False)\n",
    "\n",
    "# edit DBTBS dataframe such that it is compatible for concatenation\n",
    "DBTBS_promoters_df_to_concat = reorganize_promoter_df_columns(DBTBS_promoters_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inducible promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_bacterial_inducible_promoters(save_csv=True):\n",
    "\n",
    "    \"\"\"\n",
    "    parse through bacterial inducible promoters obtained from Meyer et al 2019 and apply QC / filtering\n",
    "    \n",
    "    inputs: boolean argument indicating whether csv files should be saved\n",
    "    \n",
    "    outputs: dataframe containing bacterial inducible promoters and metadata,\n",
    "            if save_csv, then csv file is saved to data directory\n",
    "    \n",
    "    \"\"\"  \n",
    "\n",
    "    # load raw inducible promoter dataframe\n",
    "    inducible_promoters_bact_df = pd.read_excel('../../data/inducible_promoters/20191202_inducible_promoters_bacteria.xlsx')\n",
    "\n",
    "    inducible_promoters_bact_df = QC_DNA_sequences(inducible_promoters_bact_df)\n",
    "\n",
    "    if save_csv:\n",
    "        inducible_promoters_bact_df.to_csv(\"../../data/parsed_promoter_data/20191129_bacterial_inducible_promoters.csv\")\n",
    "        \n",
    "    \n",
    "    return(inducible_promoters_bact_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run bacterial inducible promoters functions\n",
    "inducible_promoters_bact_df = parse_bacterial_inducible_promoters(save_csv=False)\n",
    "inducible_promoters_bact_df_to_concat = reorganize_promoter_df_columns(inducible_promoters_bact_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fungal_inducible_promoters(save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    parse through fungal inducible promoters obtained from various sources and apply QC / filtering\n",
    "    \n",
    "    inputs: boolean argument indicating whether csv files should be saved\n",
    "    \n",
    "    outputs: dataframe containing DBTBS promoters and metadata,\n",
    "            if save_csv, then csv file is saved to data directory\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # load raw inducible promoter dataframe\n",
    "    inducible_promoters_fungus_df = pd.read_excel('../../data/inducible_promoters/20191210_inducible_promoters_fungus.xlsx')\n",
    "    \n",
    "    inducible_promoters_fungus_df = QC_DNA_sequences(inducible_promoters_fungus_df)\n",
    "\n",
    "    if save_csv:\n",
    "        inducible_promoters_fungus_df.to_csv(\"../../data/parsed_promoter_data/20200128_fungal_inducible_promoter_sequences.csv\")\n",
    "\n",
    "    return(inducible_promoters_fungus_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run fungal inducible promoters functions\n",
    "inducible_promoters_fungus_df = parse_fungal_inducible_promoters(save_csv=False)\n",
    "inducible_promoters_fungus_df_to_concat = reorganize_promoter_df_columns(inducible_promoters_fungus_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def inducible_promoters_count(inducible_promoters_bact_df_to_concat, inducible_promoters_fungus_df_to_concat, save_csv=False):\n",
    "    \n",
    "    \"\"\"\n",
    "    stratify inducible promoters by organism (e.g., fungal, bacterial, mammalian)\n",
    "    \n",
    "    inputs: dataframes for bacterial and fungal inducible promoters\n",
    "    \n",
    "    outputs: dataframe containing inducible promoter counts\n",
    "    \"\"\"\n",
    "\n",
    "    bact_inducible_promoters = list(inducible_promoters_bact_df_to_concat['promoter'])\n",
    "    bact_inducible_promoter_varients_count = 0\n",
    "\n",
    "    # counting promoter varients\n",
    "    for induc_prom in bact_inducible_promoters:\n",
    "        if 'varient' in induc_prom:\n",
    "            bact_inducible_promoter_varients_count += 1\n",
    "\n",
    "    fungus_inducible_promoter_varients_count = 0\n",
    "\n",
    "    mammalian_inducible_promoter_varients_count = 0\n",
    "\n",
    "\n",
    "    inducible_promoters_counts_df = pd.DataFrame(columns = [\"Promoters total\", \"Original promoters\", \"Varients\"])\n",
    "    inducible_promoters_counts_df.loc['Mammalian', 'Promoters total'] = 0\n",
    "    inducible_promoters_counts_df.loc['Bacteria', 'Promoters total'] = inducible_promoters_bact_df_to_concat.shape[0]\n",
    "    inducible_promoters_counts_df.loc['Fungi', 'Promoters total'] = inducible_promoters_fungus_df_to_concat.shape[0]\n",
    "\n",
    "    inducible_promoters_counts_df.loc['Mammalian', 'Original promoters'] = 0 - mammalian_inducible_promoter_varients_count\n",
    "    inducible_promoters_counts_df.loc['Bacteria', 'Original promoters'] = inducible_promoters_bact_df_to_concat.shape[0] - bact_inducible_promoter_varients_count\n",
    "    inducible_promoters_counts_df.loc['Fungi', 'Original promoters'] = inducible_promoters_fungus_df_to_concat.shape[0] - fungus_inducible_promoter_varients_count\n",
    "\n",
    "    inducible_promoters_counts_df.loc['Mammalian', 'Varients'] = mammalian_inducible_promoter_varients_count\n",
    "    inducible_promoters_counts_df.loc['Bacteria', 'Varients'] = bact_inducible_promoter_varients_count\n",
    "    inducible_promoters_counts_df.loc['Fungi', 'Varients'] = fungus_inducible_promoter_varients_count\n",
    "\n",
    "    inducible_promoters_counts_df = inducible_promoters_counts_df.sort_values(by=\"Promoters total\")\n",
    "    inducible_promoters_counts_df = inducible_promoters_counts_df.astype('int')\n",
    "    \n",
    "    if save_csv:\n",
    "        inducible_promoters_counts_df.to_csv('../../data/parsed_promoter_data/20200128_inducible_promoter_counts.csv')\n",
    "    return(inducible_promoters_counts_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_inducible_promoters(inducible_promoters_counts_df, save_plot=True):\n",
    "\n",
    "    \"\"\"\n",
    "    create bar plot showing DBTBS promoters stratified by cognate sigma factor\n",
    "    \n",
    "    # inputs: dataframe containing DBTBS organisms x motifs (the output of DBTBS_motifs function)\n",
    "    \n",
    "    # outputs: if save_plot, then a bar plot is saved\n",
    "    \"\"\"\n",
    "    \n",
    "    black = '#000000'\n",
    "    dark_slate_gray = '#2F4F4F'\n",
    "    plot_colors = [black, dark_slate_gray]\n",
    "\n",
    "\n",
    "    inducible_promoters_counts_df.iloc[:,1:].plot.barh(stacked=True, color = plot_colors, figsize=(10,7))\n",
    "    plt.xlabel('Count', size=24,fontname=\"Arial\")\n",
    "    plt.ylabel('Organism', size=24,fontname=\"Arial\")\n",
    "    plt.title('Inducible promoter counts',fontname=\"Arial\", size=24)\n",
    "    plt.xticks(fontsize = 16,fontname=\"Arial\") \n",
    "    plt.yticks(fontsize = 16,fontname=\"Arial\")\n",
    "    plt.legend(loc='lower right', prop={'size': 16})\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if save_plot:\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191210_inducible_promoter_counts.pdf')\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191210_inducible_promoter_counts.png', dpi=400)\n",
    "        plt.savefig('../../figs/promoter_database_summaries/20191210_inducible_promoter_counts.svg')\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run functions for inducible promoter counts and plot these counts\n",
    "inducible_promoters_counts_df = inducible_promoters_count(inducible_promoters_bact_df_to_concat, inducible_promoters_fungus_df_to_concat)\n",
    "plot_inducible_promoters(inducible_promoters_counts_df, save_plot=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate promoters\n",
    "promoters_df = pd.concat([EPDnew_promoters_df_to_concat, \n",
    "           regulonDB_promoters_df_to_concat,\n",
    "           DBTBS_promoters_df_to_concat,\n",
    "           inducible_promoters_bact_df_to_concat,\n",
    "            inducible_promoters_fungus_df_to_concat], \n",
    "            sort=False).reset_index().drop('index', axis=1)\n",
    "\n",
    "promoters_df.to_csv(\"../../data/parsed_promoter_data/20191203_promoters.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
